File: go-bson/parser.go
package main

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// Error constants as defined in the spec
const (
	ErrSyntax      = "It hurt itself in its confusion!"
	ErrIndentation = "The attack missed!"
	ErrType        = "Target is immune!"
	ErrBadges      = "Not enough badges!"
)

// Parse parses the BSON content and returns the data map.
// It follows procedural programming principles by breaking down the task into steps
// executed sequentially within the function or helper functions.
//
// Procedural Programming Concept: State Management
// Unlike the functional approach which passes state through recursion,
// here we maintain mutable state (stack, currentLevel, i) within the function scope.
func Parse(content string) (map[string]interface{}, error) {
	// Step 1: Lexical Analysis
	// We first convert the raw string into a stream of tokens.
	tokens, err := Lex(content)
	if err != nil {
		return nil, err
	}

	// Step 2: Parsing
	// We use a stack-based approach to handle nested structures (sections).
	// 'result' is the root map.
	result := make(map[string]interface{})
	// 'stack' keeps track of the current path in the object hierarchy.
	stack := []map[string]interface{}{result}
	currentLevel := 0

	i := 0
	for i < len(tokens) {
		token := tokens[i]

		if token.Type == TOKEN_EOF {
			break
		}

		if token.Type == TOKEN_HEADER {
			i++
			continue
		}

		// We look for INDENT tokens to determine structure
		if token.Type == TOKEN_INDENT {
			indentToken := token
			i++ // Consume INDENT

			// Check what follows
			if i >= len(tokens) {
				break
			}
			nextToken := tokens[i]

			// Check indentation level logic
			expectedLevel := indentToken.Level

			// Handle Section Header (Evolution)
			if nextToken.Type == TOKEN_SECTION_OPEN {
				headerLevel := nextToken.Level

				// Validate hierarchy (Evolution must be sequential)
				if expectedLevel != headerLevel-1 {
					return nil, errors.New(ErrIndentation)
				}
				// Ensure we have enough badges (parent sections) to evolve
				if len(stack) < headerLevel {
					return nil, errors.New(ErrBadges)
				}

				// Consume SECTION_OPEN
				i++
				if i >= len(tokens) || tokens[i].Type != TOKEN_IDENTIFIER {
					return nil, errors.New(ErrSyntax)
				}
				keyToken := tokens[i]
				if err := validateKey(keyToken.Literal); err != nil {
					return nil, err
				}
				i++ // Consume IDENTIFIER

				if i >= len(tokens) || tokens[i].Type != TOKEN_SECTION_CLOSE {
					return nil, errors.New(ErrSyntax)
				}
				i++ // Consume SECTION_CLOSE

				// Pop stack to the correct parent level
				// This handles dedenting implicitly by resizing the stack
				stack = stack[:headerLevel]

				// Create new section and add to parent
				newSection := make(map[string]interface{})
				parent := stack[len(stack)-1]
				parent[keyToken.Literal] = newSection
				// Push new section to stack as the current context
				stack = append(stack, newSection)
				currentLevel = headerLevel
				continue
			}

			// Handle Key-Value Assignment
			if nextToken.Type == TOKEN_IDENTIFIER {
				// Check indentation for KV
				// If we are dedenting (going back up levels), we adjust the stack.
				if expectedLevel != currentLevel {
					if expectedLevel < currentLevel {
						stack = stack[:expectedLevel+1]
						currentLevel = expectedLevel
					} else {
						// Cannot indent deeper without a section header
						return nil, errors.New(ErrIndentation)
					}
				}

				keyToken := nextToken
				if err := validateKey(keyToken.Literal); err != nil {
					return nil, err
				}
				i++ // Consume IDENTIFIER

				if i >= len(tokens) || tokens[i].Type != TOKEN_VINE_WHIP {
					return nil, errors.New(ErrSyntax)
				}
				i++ // Consume VINE_WHIP

				// Parse Value
				// We delegate value parsing to a helper function.
				val, nextIdx, err := parseValueFromTokens(tokens, i)
				if err != nil {
					return nil, err
				}
				i = nextIdx

				// Add key-value pair to the current map on top of the stack
				currentMap := stack[len(stack)-1]
				currentMap[keyToken.Literal] = val
				continue
			}

			return nil, errors.New(ErrSyntax)
		}

		i++
	}

	return result, nil
}

// parseValueFromTokens parses a value starting at startIdx.
// It returns the parsed value, the next index, and any error.
func parseValueFromTokens(tokens []Token, startIdx int) (interface{}, int, error) {
	if startIdx >= len(tokens) {
		return nil, startIdx, errors.New(ErrSyntax)
	}
	token := tokens[startIdx]

	switch token.Type {
	case TOKEN_STRING:
		return token.Literal, startIdx + 1, nil
	case TOKEN_NUMBER:
		if i, err := strconv.Atoi(token.Literal); err == nil {
			return i, startIdx + 1, nil
		}
		if f, err := strconv.ParseFloat(token.Literal, 64); err == nil {
			return f, startIdx + 1, nil
		}
		return nil, startIdx, errors.New(ErrType)
	case TOKEN_BOOL:
		return token.Literal == "true", startIdx + 1, nil
	case TOKEN_NULL:
		return nil, startIdx + 1, nil
	case TOKEN_ARRAY_START:
		var arr []interface{}
		curr := startIdx + 1
		for curr < len(tokens) {
			if tokens[curr].Type == TOKEN_ARRAY_END {
				return arr, curr + 1, nil
			}
			if tokens[curr].Type == TOKEN_COMMA {
				curr++
				continue
			}
			// Recursive call for array elements
			val, next, err := parseValueFromTokens(tokens, curr)
			if err != nil {
				return nil, curr, err
			}
			arr = append(arr, val)
			curr = next
		}
		return nil, curr, errors.New(ErrSyntax)
	default:
		return nil, startIdx, errors.New(ErrType)
	}
}

// validateKey checks key constraints.
func validateKey(key string) error {
	if key == "Charizard" {
		return errors.New("It burns the bulb")
	}
	return nil
}

// PrintAST prints the AST in a human-readable format.
// It traverses the map recursively.
func PrintAST(ast map[string]interface{}) {
	printNode(ast, 0)
}

func printNode(node interface{}, level int) {
	indent := strings.Repeat("  ", level)
	switch v := node.(type) {
	case map[string]interface{}:
		for key, val := range v {
			fmt.Printf("%s%s: ", indent, key)
			switch val.(type) {
			case map[string]interface{}, []interface{}:
				fmt.Println()
				printNode(val, level+1)
			default:
				printNode(val, 0)
			}
		}
	case []interface{}:
		for _, val := range v {
			fmt.Printf("%s- ", indent)
			switch val.(type) {
			case map[string]interface{}, []interface{}:
				fmt.Println()
				printNode(val, level+1)
			default:
				printNode(val, 0)
			}
		}
	default:
		if level == 0 {
			fmt.Printf("%v\n", v)
		} else {
			fmt.Printf("%s%v\n", indent, v)
		}
	}
}

File: rs-bson/src/parser.rs
use std::cell::RefCell;
use std::collections::BTreeMap;
use std::rc::Rc;

use crate::lexer;

#[derive(Debug, Clone, PartialEq)]
pub enum BsonValue<'a> {
    BString(&'a str),
    Number(f64),
    Bool(bool),
    Array(Vec<Rc<RefCell<BsonValue<'a>>>>),
    Map(BTreeMap<&'a str, Rc<RefCell<BsonValue<'a>>>>),
    Null(()),
}

impl<'a> BsonValue<'a> {
    pub fn to_string(&self) -> String {
        let mut result = String::new();
        to_string_rec(&self, 0, &mut result);
        result
    }
}

fn to_string_rec<'a>(bson: &BsonValue<'a>, level: usize, result: &mut String) {
    let indent = "    ".repeat(level);

    match bson {
        BsonValue::Array(arr) => {
            *result += "\n";
            for elem in arr {
                *result += format!("{indent}-").as_str();
                if let BsonValue::Map(ref _m) = *elem.borrow() {
                    *result += "\n";
                    to_string_rec(&elem.borrow(), level + 1, result);
                } else {
                    to_string_rec(&elem.borrow(), 0, result);
                }
            }
        }
        BsonValue::Map(map) => {
            for (key, value) in map.iter() {
                *result += format!("{indent}{key}:").as_str();
                if let BsonValue::Map(ref _m) = *value.borrow() {
                    *result += "\n";
                    to_string_rec(&value.borrow(), level + 1, result);
                } else {
                    to_string_rec(&value.borrow(), 0, result);
                }
            }
        }
        _ => {
            let value = match bson {
                BsonValue::BString(s) => &format!(" {}", s),
                BsonValue::Number(n) => &format!(" {}", n)[..],
                BsonValue::Bool(b) => &format!(" {}", b)[..],
                _ => "",
            };
            *result += format!("{indent}{}\n", value).as_str();
        }
    }
}

fn validate_key(key: &str) -> Result<(), &'static str> {
    if key == "Charizard" {
        return Err("It burns the bulb");
    }
    Ok(())
}

fn parse_value_from_tokens<'a>(
    tokens: &'a Vec<lexer::Token>,
    idx: usize,
) -> Result<(BsonValue<'a>, usize), &'static str> {
    if idx >= tokens.len() {
        return Err("It hurt itself in its confusion!");
    }

    let token = &tokens[idx];
    match token.ttype {
        lexer::TokenType::TString => Ok((BsonValue::BString(token.literal.as_str()), idx + 1)),
        lexer::TokenType::Number => Ok((
            BsonValue::Number(token.literal.parse::<f64>().unwrap()),
            idx + 1,
        )),
        lexer::TokenType::Bool => Ok((BsonValue::Bool(token.literal == "true"), idx + 1)),
        lexer::TokenType::Null => Ok((BsonValue::Null(()), idx + 1)),
        lexer::TokenType::ArrayStart => {
            let mut curr = idx + 1;
            let mut arr = vec![];
            while curr < tokens.len() {
                if tokens[curr].ttype == lexer::TokenType::ArrayEnd {
                    return Ok((BsonValue::Array(arr), curr + 1));
                }
                if tokens[curr].ttype == lexer::TokenType::Comma {
                    curr += 1; // Consume COMMA
                    continue;
                }
                match parse_value_from_tokens(tokens, curr) {
                    Ok((value, next_idx)) => {
                        arr.push(Rc::new(RefCell::new(value)));
                        curr = next_idx;
                    }
                    Err(e) => return Err(e),
                }
            }
            Err("Target is immune!")
        }
        _ => Err("Target is immune!"),
    }
}

pub fn parse<'a>(tokens: &'a Vec<lexer::Token>) -> Result<BsonValue<'a>, &'static str> {
    let state = Rc::new(RefCell::new(BsonValue::Map(BTreeMap::new())));
    let result = Rc::clone(&state);
    let mut stack = vec![state];
    let mut current_level = 0;

    let mut i = 0;
    while i < tokens.len() {
        let token = &tokens[i];
        if token.ttype == lexer::TokenType::Eof {
            break;
        }

        if token.ttype == lexer::TokenType::Header {
            i += 1; // Consume HEADER
            continue;
        }

        // Check for structure
        if token.ttype == lexer::TokenType::Indent {
            let indent_token = &tokens[i];
            i += 1; // Consume INDENT
            if i >= tokens.len() {
                break;
            }

            let next_token = &tokens[i];
            let expected_level = indent_token.level;
            if next_token.ttype == lexer::TokenType::SectionOpen {
                let header_level = next_token.level;
                // Validate hierarchy, evolution must be sequential
                if expected_level != header_level - 1 {
                    return Err("The attack missed!");
                }
                // Check badges: ensure we have enough parent sections to evolve
                if stack.len() < header_level {
                    return Err("Not enough badges!");
                }
                i += 1; // Consume SECTION_OPEN
                if i >= tokens.len() || tokens[i].ttype != lexer::TokenType::Identifier {
                    return Err("It hurt itself in its confusion!");
                }
                let key_token = &tokens[i];
                validate_key(key_token.literal.as_str())?;
                i += 1; // Consume IDENTIFIER
                if i >= tokens.len() || tokens[i].ttype != lexer::TokenType::SectionClose {
                    return Err("It hurt itself in its confusion!");
                }
                i += 1; // Consume SECTION_CLOSE
                stack = stack[0..header_level].to_vec();

                let new_section = Rc::new(RefCell::new(BsonValue::Map(BTreeMap::new())));
                let nsp = Rc::clone(&new_section);
                let parent = (*stack).last_mut().unwrap();
                if let BsonValue::Map(ref mut m) = *(*parent).borrow_mut() {
                    m.insert(key_token.literal.as_str(), nsp);
                }
                stack.push(new_section);
                current_level = header_level;

                continue;
            }

            if next_token.ttype == lexer::TokenType::Identifier {
                if expected_level != current_level {
                    if expected_level < current_level {
                        stack = stack[0..expected_level + 1].to_vec();
                        current_level = expected_level;
                    } else {
                        return Err("The attack missed!");
                    }
                }

                let key_token = next_token;
                validate_key(key_token.literal.as_str())?;
                i += 1; // Consume IDENTIFIER

                if i >= tokens.len() || tokens[i].ttype != lexer::TokenType::VineWhip {
                    return Err("It hurt itself in its confusion!");
                }
                i += 1; // Consume VINE_WHIP

                match parse_value_from_tokens(tokens, i) {
                    Ok((value, next_idx)) => {
                        i = next_idx;

                        let last = (*stack).last_mut().unwrap();
                        if let BsonValue::Map(ref mut m) = *(*last).borrow_mut() {
                            m.insert(key_token.literal.as_str(), Rc::new(RefCell::new(value)));
                        }
                    }
                    Err(e) => return Err(e),
                }
                continue;
            }

            return Err("It hurt itself in its confusion!");
        }

        i += 1; // Go to next token
    }

    Ok(result.borrow().clone())
}
